File: phase0/evri/sentences.py
"""Merge fragmented ASR cues into whole sentences.

This is the step that makes the difference between our output and YouTube's own
Turkish auto-translation. See DESIGN.md section 4.

ASR captions frequently have NO punctuation at all, so punctuation alone is not
a sufficient boundary signal — the gap, duration and length limits carry most of
the weight in practice.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from .cues import Cue

SENTENCE_END_RE = re.compile(r"[.!?…]['\")\]]*$")


@dataclass
class Sentence:
    start_ms: int
    end_ms: int
    text: str
    cue_count: int


def merge_into_sentences(
    cues: list[Cue],
    *,
    max_gap_ms: int = 1200,
    max_duration_ms: int = 6000,
    # ASR input often has no punctuation at all, so this cap — not the full stop —
    # is what actually ends most groups. At 200 a Turkish line renders as 3-4 rows
    # of subtitle; 120 keeps it to the two rows a TV viewer can read in time.
    max_chars: int = 120,
) -> list[Sentence]:
    """Group cues into sentence-ish units.

    A group is closed when any of these is true:
      * the accumulated text ends with sentence-final punctuation
      * the next cue starts more than ``max_gap_ms`` after this one ends
      * the group has reached ``max_duration_ms`` or ``max_chars``
    """
    sentences: list[Sentence] = []
    buffer: list[Cue] = []

    def flush() -> None:
        if not buffer:
            return
        text = " ".join(cue.text for cue in buffer)
        text = re.sub(r"\s+", " ", text).strip()
        if text:
            sentences.append(
                Sentence(buffer[0].start_ms, buffer[-1].end_ms, text, len(buffer))
            )
        buffer.clear()

    for index, cue in enumerate(cues):
        buffer.append(cue)

        text = " ".join(item.text for item in buffer).strip()
        duration = buffer[-1].end_ms - buffer[0].start_ms

        gap_to_next = None
        if index + 1 < len(cues):
            gap_to_next = cues[index + 1].start_ms - cue.end_ms

        if SENTENCE_END_RE.search(text):
            flush()
        elif duration >= max_duration_ms or len(text) >= max_chars:
            flush()
        elif gap_to_next is not None and gap_to_next > max_gap_ms:
            flush()

    flush()
    return sentences


def sentences_to_cues(sentences: list[Sentence], translations: list[str]) -> list[Cue]:
    """Pair translated text back onto the source sentences' time ranges.

    We deliberately do NOT try to redistribute a translation across the original
    fragment cues: Turkish word order differs from English, so any such split
    would put words under the wrong timestamps.
    """
    if len(sentences) != len(translations):
        raise ValueError(
            f"sentence/translation count mismatch: "
            f"{len(sentences)} vs {len(translations)}"
        )
    return [
        Cue(sentence.start_ms, sentence.end_ms, text.strip())
        for sentence, text in zip(sentences, translations)
        if text.strip()
    ]

File: app/src/main/java/tr/com/uslanozan/evritext/subtitles/Sentences.kt
package tr.com.uslanozan.evritext.subtitles

/**
 * Merge fragmented ASR cues into whole sentences, following the prototype's measured rules.
 *
 * This is the step that makes the difference between our output and YouTube's own
 * Turkish auto-translation (DESIGN.md section 4): translating two-second fragments
 * produces nonsense, translating sentences does not.
 *
 * ASR captions frequently carry no punctuation at all, so the full stop is *not* the
 * boundary that does the work in practice — the character and duration caps are.
 */
data class Sentence(
    val startMs: Long,
    val endMs: Long,
    val text: String,
    val cueCount: Int,
    /** Original ASR timing anchors, retained for target-language resegmentation. */
    val sourceCues: List<Cue> = emptyList(),
)

object Sentences {

    private val SENTENCE_END = Regex("""[.!?…]['")\]]*$""")
    private val WHITESPACE = Regex("""\s+""")

    /**
     * A group closes when any of these is true:
     *  * the accumulated text ends with sentence-final punctuation
     *  * the next cue starts more than [maxGapMs] after this one ends
     *  * the group has reached [maxDurationMs] or [maxChars]
     *
     * [maxChars] is 120 rather than the 200 Phase 0 started with: at 200 a Turkish
     * line renders as three or four rows of subtitle, which is more than a viewer can
     * read before the cue changes.
     */
    fun merge(
        cues: List<Cue>,
        maxGapMs: Long = 1200,
        maxDurationMs: Long = 6000,
        maxChars: Int = 120,
    ): List<Sentence> {
        val sentences = mutableListOf<Sentence>()
        val buffer = mutableListOf<Cue>()

        fun flush() {
            if (buffer.isEmpty()) return
            val text = WHITESPACE.replace(buffer.joinToString(" ") { it.text }, " ").trim()
            if (text.isNotEmpty()) {
                sentences.add(
                    Sentence(
                        startMs = buffer.first().startMs,
                        endMs = buffer.last().endMs,
                        text = text,
                        cueCount = buffer.size,
                        sourceCues = buffer.toList(),
                    ),
                )
            }
            buffer.clear()
        }

        cues.forEachIndexed { index, cue ->
            buffer.add(cue)

            val text = buffer.joinToString(" ") { it.text }.trim()
            val duration = buffer.last().endMs - buffer.first().startMs
            val gapToNext = cues.getOrNull(index + 1)?.let { it.startMs - cue.endMs }

            when {
                SENTENCE_END.containsMatchIn(text) -> flush()
                duration >= maxDurationMs || text.length >= maxChars -> flush()
                gapToNext != null && gapToNext > maxGapMs -> flush()
            }
        }
        flush()
        return sentences
    }

    /**
     * Pairs translated text back onto the source sentences' time ranges.
     *
     * Translation still happens over the complete sentence. Only after that do we
     * split obviously long target text at semantic boundaries, snapping each split to
     * an original ASR timing anchor. Unsafe cases retain the original single cue.
     */
    fun toCues(sentences: List<Sentence>, translations: List<String>): List<Cue> {
        require(sentences.size == translations.size) {
            "sentence/translation count mismatch: ${sentences.size} vs ${translations.size}"
        }
        return sentences.zip(translations).flatMap { (sentence, text) ->
            TargetCueSegmenter.segment(sentence, text)
        }
    }
}
